include/openssl/params.h

```diff
@@ -0,0 +1,180 @@
+# define OSSL_PARAM_SIZED_double(key, addr, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_REAL, (addr), sizeof(double), &(r_sz))
+
+# define OSSL_PARAM_SIZED_BN(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UNSIGNED_INTEGER, (addr), sz, \
+                    &(r_sz))
+
+# define OSSL_PARAM_SIZED_utf8_string(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UTF8_STRING, (addr), sz, &(r_sz))
+# define OSSL_PARAM_SIZED_octet_string(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_OCTET_STRING, (addr), sz, &(r_sz))
+
+# define OSSL_PARAM_SIZED_utf8_ptr(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_UTF8_PTR, &(addr), sz, &(r_sz))
+# define OSSL_PARAM_SIZED_octet_ptr(key, addr, sz, r_sz) \
+    OSSL_PARAM_DEFN((key), OSSL_PARAM_OCTET_PTR, &(addr), sz, &(r_sz))
+
+/* Search an OSSL_PARAM array for a matching name */
+const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key);
+
+/* Basic parameter type run-time construction */
+OSSL_PARAM OSSL_PARAM_construct_int(const char *key, int *buf, size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_uint(const char *key, unsigned int *buf,
+                                     size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_long(const char *key, long int *buf,
+                                     size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_ulong(const char *key, unsigned long int *buf,
+                                     size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_int32(const char *key, int32_t *buf,
+                                      size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_uint32(const char *key, uint32_t *buf,
+                                      size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_int64(const char *key, int64_t *buf,
+                                      size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_uint64(const char *key, uint64_t *buf,
+                                       size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_size_t(const char *key, size_t *buf,
+                                       size_t *ret);
+OSSL_PARAM OSSL_PARAM_construct_BN(const char *key, unsigned char *buf,
+                                   size_t bsize, size_t *rsize);
+OSSL_PARAM OSSL_PARAM_construct_double(const char *key, double *buf,
+                                       size_t *rsize);
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, char *buf,
+                                            size_t bsize, size_t *rsize);
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, char **buf,
+                                                size_t bsize, size_t *rsize);
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize);
+OSSL_PARAM OSSL_PARAM_construct_octet_string_ptr(const char *key, void **buf,
+                                                 size_t bsize, size_t *rsize);
```

Comment:
Shouldn't `OSSL_PARAM_construct_utf8_string_ptr` and `OSSL_PARAM_construct_octet_string_ptr` be renamed to `OSSL_PARAM_construct_utf8_ptr` and `OSSL_PARAM_construct_octet_ptr`, for consistency?

Reply:
Yeah, I found this independently.  I've pushed the update.